solace/src/main/java/org/apache/apex/malhar/solace/AbstractSolaceGuaranteedIdempotentInputOperator.java

```diff
@@ -0,0 +1,649 @@
+import com.solacesystems.jcsmp.Endpoint;
+import com.solacesystems.jcsmp.EndpointProperties;
+import com.solacesystems.jcsmp.FlowReceiver;
+import com.solacesystems.jcsmp.JCSMPException;
+import com.solacesystems.jcsmp.JCSMPProperties;
+import com.solacesystems.jcsmp.XMLMessageListener;
+import com.datatorrent.api.Attribute;
+import com.datatorrent.api.Context;
+import com.datatorrent.api.Context.DAGContext;
+import com.datatorrent.api.Context.OperatorContext;
+import com.datatorrent.api.InputOperator;
+import com.datatorrent.api.Operator;
+import com.datatorrent.api.Partitioner;
+import com.datatorrent.api.annotation.OperatorAnnotation;
+import com.datatorrent.common.partitioner.StatelessPartitioner;
+import com.datatorrent.lib.io.IdempotentStorageManager;
+import com.datatorrent.netlet.util.DTThrowable;
+
+@OperatorAnnotation(checkpointableWithinAppWindow = false)
+public abstract class AbstractSolaceGuaranteedIdempotentInputOperator<T> extends AbstractSolaceBaseInputOperator<T> implements InputOperator, Operator.ActivationListener<Context.OperatorContext>
+{
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractSolaceGuaranteedIdempotentInputOperator.class);
+
+  @NotNull
+  protected String endpointName;
+
+
+  @NotNull
+  protected transient EndpointType endpointType = EndpointType.QUEUE;
+  private transient Endpoint endpoint;
+  @NotNull
+  public transient FlowCallbackMessageHandler FlowHandler = new FlowCallbackMessageHandler();
+  @NotNull
+  protected transient EndpointProperties endpointProperties = new EndpointProperties();
+  private transient BytesXMLMessage recentMessage = null;
+
+  private transient long[] operatorRecoveredWindows;
+  protected transient long currentWindowId;
+  @SuppressWarnings("unused")
+  private final transient AtomicReference<Throwable> throwable;
+  protected transient ArrayBlockingQueue<BytesXMLMessage> arrivedMessagesToProcess;
+  protected transient BlockingQueue<BytesXMLMessage> unackedMessages;
+
+  protected transient TreeMap<Long, BytesXMLMessage> lastMessages = new TreeMap<Long, BytesXMLMessage>();
+  @SuppressWarnings("unused")
+  private transient long windowTime;
+  protected final transient Map<Long, T> currentWindowRecoveryState;
+  protected transient LinkedList<T> inFlightRecoveryMessages = new LinkedList<T>(); // used by partition that restarts for duplicate detection
+  protected transient LinkedList<T> inFlightRecoveryMessagesPartition = new LinkedList<T>(); // used by partitions that didn't restart for duplicate detection
+  protected transient LinkedList<T> inFlightRecoveryMessagesDR = new LinkedList<T>(); // used by partition that detect DR fail over for duplicate detection
+
+  private transient Context.OperatorContext context;
+  private transient int partitionCount = 0;
+
+  protected static final transient int DEFAULT_BUFFER_SIZE = 500;
+  protected static transient int DR_COUNTER_SIZE = 2 * DEFAULT_BUFFER_SIZE;
+  protected transient int DR_COUNTER = 0;
```

Comment:
Why are the above 2 variables defined in upper case when they are not constants? Can we change to camel-case? Also DR_COUNTER_SIZE is class static but I am not sure if that is intended.